### PR TITLE
Feature/models

### LINKED ISSUE #65: implement error checking when adding layers to Variational()
see https://github.com/blei-lab/edward/pull/62#issuecomment-217540682

### LINKED ISSUE #166: testing for Variational
+ `models/test_variational_add.py`
+ `models/test_variational_init.py`
+ `models/test_variational_log_prob.py`
+ `models/test_variational_sample.py`

### PR DESCRIPTION
**Pull request by dustinvtran**
#### Summary:

+ issues: related to #7; related to #27; fix #65 (by removing `Variational` entirely); related to #66; related to #154; fix #166 (by removing `Variational` entirely)

Interfacing with multiple sets of latent variables is unintuitive in Edward. For one set of latent variables, a model's `log_prob(xs, zs)` accesses `zs` as a tf.Tensor. For multiple sets of latent variables, `zs` is a list;  `log_prob(xs, zs)` accesses a specific ordering of `zs` according to how the variational model is defined.

This pull request modifies the API of all components of Edward in order to make multiple sets of latent variables more intuitive. These changes enable fine-grained statements about random variables (that is, locally per random variable rather than globally over the entire collection). This in turn enables more fine-grained experimentation.

__API for model methods which calculate a density.__
Recall that data is a dictionary, and `log_prob(xs, zs)` grabs values of `xs` according to user-defined keys. `zs` takes the same approach: `zs` is now a dictionary, and `log_prob(xs, zs)` grabs values of `zs` according to user-defined keys. Here's an example:
```python
class BetaBernoulli:
    """p(x, p) = Bernoulli(x | p) * Beta(p | 1, 1)"""
    def log_prob(self, xs, zs):
        log_prior = beta.logpdf(zs['p'], a=1.0, b=1.0)
        log_lik = tf.pack([tf.reduce_sum(bernoulli.logpmf(xs['x'], p))
                           for p in tf.unpack(zs['p'])])
        return log_lik + log_prior
```

__API for inference.__
The change to density methods is obvious. To get this to work, we have to pass in these strings during inference. Inference is modified so that users must pass in `latent_vars` as the first argument. It is always a dictionary binding each string to the corresponding posterior factor.
```python
ed.MAP({'z': qz}, data, model)
ed.MCMC({'pi': qpi, 'mu': qmu, 'sigma': qsigma}, data, model)
ed.VariationalInference({'z': qz}, data, model)
```
(You may wonder why the ordering is `latent_vars`, `data`, `model`, and in fact why internally, `model` is written as `model_wrapper`. This will become obvious once we add our modeling language, where `model_wrapper` is an optional argument.)

__API for variational models.__
With the above inference API, variational models no longer need a container object. Everything is defined locally with respect to each variational factor. For example,
```python
qpi = Dirichlet(K)
qmu = Normal(K*D)
qsigma = Normal(K*D)
inference = ed.MFVI({'pi': qpi, 'mu': qmu, 'sigma': qsigma}, data, model)
```
Defining things locally also lets us develop more complicated inference algorithms that work on a per-variational factor basis, e.g., #66. 

Note that even MAP and MCMC require explicitly defining the inferred latent variable factors, with `PointMass()` and `Empirical()` distributions correspondingly. Practically this is useful so that we can work with these inferred factors after inference, where inference only modifies the model in-place and does not output anything. In principle this is useful for reasons described later in the modeling language paper.

__API for criticism.__
This is analogous to the changes to inference. Unlike inference, the order of `data` is before `latent_vars`.
```python
# log-likelihood performance
ed.evaluate('log_likelihood', data={'x': x_train}, latent_vars={'z': qz}, model_wrapper=model)
# classification accuracy
ed.evaluate('binary_accuracy', data={'y': y_train, 'x': x_train}, latent_vars={'z': qz}, model_wrapper=model)
# posterior predictive check
ed.ppc(T, data={'x': x_train}, latent_vars={'z': qz}, model_wrapper=model)
# prior predictive check
ed.ppc(T, data={'x': x_train}, latent_vars=None, model_wrapper=model)
```

Footnotes
 + The implementation of certain loss functions is inefficient. For example, in several loss functions during `MFVI`, I first loop over `latent_vars` to create a dictionary of samples. Then I loop over it again to calculate the variational distribution's log-density for each sample. This is purposefully done for readability. This is only inefficient during construction time; so it's negligible in the grand scheme of computational things during runtime.

+ Due to internal reasons, a list of >1 element in MAP is not supported, e.g.,
  ```
  ed.MAP(['pi', 'mu', 'sigma'], data, model_wrapper)
  ```
  See doc for why. Or the MAP examples.
+ The PyMC3 and Stan models are sped up. This is because internally, model wrappers no longer do a flattened vector-to-dictionary conversion for the latent variables.

todo's before merging
+ [x] update `criticisms.py`
+ [x] update `models/models.py` | model wrappers to enable latent variable dictionary
+ [x] update `docs/tex/` | getting started, delving in, inference tutorials, etc.
+ [x] update `docs/source/` | inference api
+ [x] update `examples/`
+ [x] update `tests/` | `test_inference_data.py`, `test-models/test_{pymc3model,pythonmodel,stanmodel}_log_prob.py`

#### Intended Effect:

See above.

#### How to Verify:

Run all examples and unit tests.

#### Side Effects:

None.

#### Documentation:

Updated the documentation to everywhere with changed code.

#### Reviewer Suggestions:

Anyone interested in seeing the future of Edward's model and inference API. :) 

Also @heywhoah because it's (probably) the last incremental pull request before our pull request with the new modeling language.

### CODE REVIEW

**Comment by mariru** [comment]
adjusted examples `bayesian_linear_regression_plot.py`, `bayesian_nn.py` and `bayesian_analytic_kl.py`

**Comment by dustinvtran** [comment]
Finally complete! Okay, time to rebase the commits.
